src/refactor_nrepl/config.clj

```diff
@@ -34,7 +34,14 @@
    :libspec-whitelist ["^cljsjs"]
 
    ;; Regexes matching paths that are to be ignored
-   :ignore-paths []})
+   :ignore-paths []
+
+   ;; Will be forwarded to clojure.pprint/*print-miser-width* when pprinting ns forms.
+   ;; You can set it to 40 for restoring traditional behavior.
+   :print-miser-width nil
+   ;; Will be forwarded to clojure.pprint/*print-right-margin* when pprinting ns forms.
+   ;; You can set it to 72 for restoring traditional behavior.
+   :print-right-margin nil})
```

Comment:
Java class names are sometimes very long.  If I'm not mistaken, the proposed change would mean that we could potentially create import clauses that would be so long nobody would be happy with them.

I don't think anyone would want that, even if I can see the argument that 72 is too restrictive.

Can we set this to 100 instead?  Having sensible defaults is very important imo, so the behavior you get out of the box is at least workable.

Reply:
The tradeoff I'm making (which reasoning I only partly shared in advance - my bad!) is that I'm fostering an agnostic/minimalistic formatting that other tools can independently, consistently achieve.

In specific terms, our impl is based on clojure.pprint. Other similar ns formatters are using rewrite-clj or vanilla clojure.

So by coupling our result to a clojure.pprint impl detail, we create a non-reproducible formatting for other tools.

`how-to-ns`, the old "standard" for this domain doesn't mention a limit of columns. So it is cautious to stick to it without introducing extra rules.

One could always add the extra newlines either with the provided new config options, or perhaps a generalized "column formatter" e.g. zprint, i.e. teams who care about this most likely already have a formatter/linter for the entire problem (.clj files, beyond ns forms).

Reply:
> The tradeoff I'm making (which reasoning I only partly shared in advance - my bad!) is that I'm fostering an agnostic/minimalistic formatting that other tools can independently, consistently achieve.

That's fine, but making the tool harder to use / have worse formatting out of the box seems like a high price to pay.  It's essentially a race to the bottom, because the easiest and most consistent way to format the ns form is to put everything on a single line that runs on for however many columns is needed...

I was going to propose moving the (imo sensible) defaults to clj-refactor instead, but that wouldn't help, right?  What you want is essentially less clever formatting, so it can be widely replicated across the tooling universe?

When I wrote `clean-ns` back in the day, what I wanted most of all was to simply pull in a library that pretty printed the ns form in a nice way.  I looked at every alternative available at the time, but couldn't find one that worked well in practice.  Perhaps your goal could be even better served if we pulled this pretty printer out into it's own library so everyone else could re-use it (with whatever config settings they want)?

> how-to-ns, the old "standard" for this domain doesn't mention a limit of columns. So it is cautious to stick to it without introducing extra rules.

FWIW, if people were consistently submitting PRs to him with import clauses that were 140 columns long, when doing java interop, I'm sure he would've mentioned it, but nobody would do that when they were manually editing the ns form.  Tooling with bad defaults might, though!

meta comment: in the grand scheme of things this isn't a big deal and I can obviously live with whatever the outcome is, but since you opened a PR I figured I would chime in and fight for ease-of-use on behalf of our users :)  Writing this all out also made me think of this old xkcd strip:

![](https://imgs.xkcd.com/comics/duty_calls.png)

Reply:
I'd disagree about what the best formatting is anyway, historically refactor-nrepl's line wrapping meant that `M-x sort-lines` wouldn't work.

But keeping things practical, we can observe that this is a flip-coin scenario so we can default to not changing things. Setting .dir-locals is easy enough.

I also particularly care about keeping our own team happy 🍻